emission/analysis/modelling/tour_model/similarity.py

```diff
@@ -1,170 +1,177 @@
-        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
-        d = R * c
-        if option==True:
-            print d
-        if d <= self.radius:
-            return True
-        return False
+# Standard imports
+import math
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+import numpy
+from sklearn import metrics
+import sys
+from scipy.optimize import curve_fit
+from numpy import cross
+from numpy.linalg import norm
+"""
+This class organizes data into bins by similarity. It then orders the bins 
+by largest to smallest and removes the bottom portion of the bins. 
+
+Two trips are in the same bin if both their start points and end points 
+are within a certain number of meters of each others. 
+
+As input, this class takes the following:
+- data: the data to put into bins. The data should be a list of Trip objects that have 
+start and end locations. 
+- radius: the radius for determining how close the start points and end points of two 
+trips have to be for the trips to be put in the same bin
+
+This is called by cluster_pipeline.py.
+"""
+class similarity:
+    
+    def __init__(self, data, radius):
+        self.data = data
+        if not data:
+            self.data = []
+        self.bins = []
+        self.radius = float(radius)
+        for a in range(len(self.data)-1, -1, -1):
+            start_lat = self.data[a].trip_start_location.lat
+            start_lon = self.data[a].trip_start_location.lon
+            end_lat = self.data[a].trip_end_location.lat
+            end_lon = self.data[a].trip_end_location.lon
+            if self.distance(start_lat, start_lon, end_lat, end_lon):
+                self.data.pop(a)
+        print 'After removing trips that are points, there are ' + str(len(self.data)) + ' data points'
+        self.size = len(self.data)
+
+    #create bins
+    def bin_data(self):
+        for a in range(self.size):
+            added = False
+            for bin in self.bins:
+                if self.match(a,bin):
+                    bin.append(a)
+                    added = True
+                    break
+            if not added:
+                self.bins.append([a])
+        self.bins.sort(key=lambda bin: len(bin), reverse=True)
+
+    #delete lower portion of bins
+    def delete_bins(self):
+        if len(self.bins) <= 1:
+            return
+        num = self.elbow_distance()
+        for i in range(len(self.bins)):
+            if len(self.bins[i]) <= len(self.bins[num]):
+                num = i
+                break
+        print 'the cutoff point is ' + str(num)
+        self.num = num
+        self.graph()
+        for i in range(len(self.bins) - num):
+            self.bins.pop()
+        newdata = []
+        for bin in self.bins:
+            for b in bin:
+                d = self.data[b]
+                newdata.append(self.data[b])
+        self.newdata = newdata
+
+    #calculate the cut-off point in the histogram
+    def elbow_distance(self):

```

Comment:
Can you provide a short writeup of the motivation here? Not as comprehensive as the paper, but a quick idea on what you are doing.